systemd: logs: Use reportd for reporting problems

ernestask opened the pull request:
The current problem reporting workflow only supports sending micro-reports to FAF (now ABRT Analytics) using a CLI utility. ABRT itself can send reports to more places: Bugzilla (Red Hat), MantisBT (CentOS), and whatever new shiny places that might be added in the future. Moreover, with ABRT, the user can send core dumps to a remote retracing server, to make stack traces in reports more useful.

Aside from not supporting all the different reporting workflows, running reporting commands directly might result in prompts being shown in the shell with no way for a remote user to respond to those.

Cue reportd, a D-Bus service that allows reporting plugins and the end user to communicate. The goal of this PR is to integrate it into Cockpit for more complete reporting capabilities.

Some basic screencasts for now:

  * [Single workflow, duplicate report found in ABRT Analytics](https://cloud.gnome.org/s/9Zciod995FepQzF)
  * [Single workflow, unique ABRT Analytics report, duplicate Bugzilla report](https://cloud.gnome.org/s/sMZBKQxyidZQ6Gq)

Some shortcoming of the implementation:

  * There is currently no way to cancel the reporting process; this is a problem on the ABRT side, and reworking the machinery for that would be somewhat involved.
  * Reporting using multiple workflows at once: probably shouldn’t be allowed (no way to distinguish between prompts, might have unexpected results), with a modal dialog or somesuch blocking interaction, also allowing to display progress information that gets sent to reportd.
  * Only a single URL to a report is accessible, which will be ABRT Analytics most of the time, but could include Bugzilla and friends as well.

ernestask commented:
P.S. reportd 0.5 is required for this, but it hasn’t reached F29 yet: https://bodhi.fedoraproject.org/updates/reportd-0.5-1.fc29

KKoukiou commented on pkg/systemd/logs.js:
```diff
@@ -399,6 +409,167 @@ $(function() {
+            case PromptType.ASK_PASSWORD: {
+                input.attr("type", "password");
+            }
+            // falls through
```
This comment is redundant. Same for all comment with same string..

ernestask replied:
It really is not:
```
/home/ekulik/Development/github.com/cockpit-project/cockpit/pkg/systemd/logs.js
  487:13  error  Expected a 'break' statement before 'case'  no-fallthrough
```

KKoukiou replied:
Oh sorry, probably it's some eslint rule we had disabled all this time.

KKoukiou commented on pkg/systemd/logs.js:
```diff
@@ -399,6 +409,167 @@ $(function() {
+            // falls through
+            case PromptType.ASK: {
+                input.attr('type', 'text');
+
```
Too many new lines here.

KKoukiou commented on pkg/systemd/logs.js:
```diff
@@ -399,6 +409,167 @@ $(function() {
+                input.attr("id", "prompt-dialog-checkbox");
+                input.attr("type", "checkbox");
+                input.css("margin-top", "1em");
+
```
What does this newline separate here?

KKoukiou commented on pkg/systemd/logs.js:
```diff
@@ -414,48 +585,95 @@ $(function() {
+                .done(function() {
+                    reportd.GetWorkflows(problem.path).done(function(args, options) {
+                        if (args.length > 1) {
+                            var menu = $('<ul class="dropdown-menu dropdown-menu-right" role="menu"></ul>');
```
There is inconsistent usage of let/var and from what I see they are used interchangeably in this patch. Can you please make these a bit consistent?

ernestask replied:
Can you provide specific instances? Such broad comments really don’t help me.

ernestask replied:
The idea is to have logical scoping, so I tried to only use `var` when the variable was needed in a callback. It might have gotten out of hand due to several rewrites, however.

KKoukiou replied:
So if you used them deliberately it's fine. I just wanted to make sure that it's not random, since I don't know your background with js :)

ernestask replied:
In that case - my background is nearly nonexistent, but most of the things were done deliberately. Maybe sometimes uninformedly, but deliberately. Some things (like vertical whitespace, code layout that reeks of C-isms) I will not try to defend, because I don’t have the energy to.

</ramble>

KKoukiou replied:
Hope these nit picks didn't discourage you. Please ignore, and please post some screenshots of the new feature + some tests to be able to do some more to the point review ;)

Thanks!

KKoukiou commented on pkg/systemd/logs.js:
```diff
@@ -21,13 +21,23 @@ import $ from "jquery";
     const _ = cockpit.gettext;
 
     var problems_client = cockpit.dbus('org.freedesktop.problems', { superuser: "try" });
+    var reportd_client = cockpit.dbus('org.freedesktop.reportd', { bus: 'session' });
```
What about root user that doesn't have session bus?

KKoukiou replied:
Better put in right before the reportd variable, instead of mixing with the problems one.

ernestask replied:
That is something I missed completely. reportd will only listen on the session bus, so that’s a bit of a problem. I can get it to work by using a bus-activatable systemd service unit, but it’s far from a good user experience, as the proxy creation fails initially, but after a refresh it’s fine.

Something to figure out still.

KKoukiou commented on pkg/systemd/logs.js:
```diff
@@ -399,6 +409,167 @@ $(function() {
+                }
+            }
+
+            return null;
```
This return is redundant.

ernestask commented:
Sorry about the barebones description. I had troubles recording a screencast, but later I will come up with a more complete package to paint the picture.

ernestask commented:
I’ve been stuck for a while chasing bugs in reportd and other places, but I believe this is in workable shape now. Updated the summary, which could use some more screencasts, but those will come later.

ernestask commented:
I should also probably add a test for multiple workflows. However, some things can simply not be tested reliably, like letting ABRT run `abrt-action-perform-ccpp-analysis`, which requires network access and can take a long time.

marusak commented:
> I should also probably add a test for multiple workflows. However, some things can simply not be tested reliably, like letting ABRT run abrt-action-perform-ccpp-analysis, which requires network access and can take a long time.

It would be much better to create some simple demo workflow and use that one.

ernestask commented:
> * When `reportd` is not present, the disabled `report` button could have on hover popup saying why it is disabled

I added a tooltip, but the need is completely negated after our most recent discussion. If I leave the current code for sending microreports, there will always be at least one “workflow”.

> * When `reportd` is installed I cannot see reported link to faf

Currently-reported (pre-ABRT 2.12/libreport 2.10/reportd 0.5) problems are indeed a problem in and of itself.

> * When I am asked for bugzilla login or password hitting enter should submit it. Now I need to click on `submit` with mouse

Yup, that’s a bad one. Fixed locally.

> * When reported the item that finished should not be present in the list - or at least when I click on it it should do something. Now when I click on it nothing happens. Also if I have there more then one item (Report to CentOs and to Fedora) neither of them on clicking does anything.
> 
> * When I start reporting process the dialog window is closed. This is rather surprising behavior, it would be better if it would not close automatically but rather let me close it (with some message that would explain that it would be running in background). Also when I click on running workflow (having spinner) I would suspect that the dialog would open so I could see progress or stop it. Also it would be nice if the spinner would be also next to the main button `Report v` so it would be obvious that something is happening.

Dropdowns are typically closed when you click on things inside, that’s just how it works. As for the other issues, I’m starting to see that using a dropdown isn’t going to work very well here. For one, reporting takes an indeterminate amount of time, during which the user might see occasional prompts for input. If the user decides to start doing other things in Cockpit while reporting is in progress, the context for those prompts will be lost (not to mention starting multiple workflows), so there is need for some kind of integration outside the package.

ernestask commented:
@marusak, pushed the latest iteration. Do note that there are a couple of failing tests:

- `testAbrtSegv`, `testAbrtDelete` (missing debugging symbols, should be fixed by https://github.com/rpm-software-management/rpm/commit/0917da26bf0722fcf40de1c5a61412f23982d131, but needs a new release of rpm and a rebuild of Python, so might take a while)
- `testAbrtReport` (https://github.com/abrt/libreport/commit/f2d98e0ee81037f18e50c06e409def19e791c2f2)

ernestask commented:
With https://bodhi.fedoraproject.org/updates/FEDORA-2019-2d7ba96ef5, `testAbrtDelete` should start passing again.

marusak commented:
This looks very nice! Thank you for working on it:)
 
I tested it and I found three possible problems:
1. On firefox when I open reported to FAF it opens two tabs with FAF instead of one. On Chrome/Chromium it works correctly.
2. I think we talked about this but I don't remeber what was the resolution - it is about possibility to cancle a task. This may be a problem when I hit by mistake different workflow and then I cannot do anything just finish it, which may be almost impossible in some cases :)
3. This one looks like a real bug, when I report with fedora it correctly  shows it was reported and upon clicking it opens BZ. However if I then open centos workflow it shows that it was already reported and points to Fedora worflow (however does not show the green tick just when I open the worflow it shows this information)

ernestask commented:
> On firefox when I open reported to FAF it opens two tabs with FAF instead of one. On Chrome/Chromium it works correctly.

It’s because of my weird setup to have the entire row be clickable and to have an anchor with `href` for accessibility. Fixed it locally, though.

>     2\. I think we talked about this but I don't remeber what was the resolution - it is about possibility to cancle a task. This may be a problem when I hit by mistake different workflow and then I cannot do anything just finish it, which may be almost impossible in some cases :)

We did talk about it. reportd hasn’t yet grown the ability to cancel tasks, so that would come a bit later, when I can figure out a good way to do it.

>     3\. This one looks like a real bug, when I report with fedora it correctly  shows it was reported and upon clicking it opens BZ. However if I then open centos workflow it shows that it was already reported and points to Fedora worflow (however does not show the green tick just when I open the worflow it shows this information)

Ugh, it’s `abrt-action-ureport` throwing up when it finds a Bugzilla report… I could probably add some sort of kludge to ignore the return code and continue with reporting.

ernestask commented:
https://bodhi.fedoraproject.org/updates/FEDORA-2019-649e4872a7 should fix your third point.

ernestask commented:
And pushed changes to address your first point.

marusak commented:
Great work!
Could you please rebase as there is conflict and write some description how to test this? (what version of which component is needed and how to make some crash). Also some video showing how this works would be very appreciated. Then please ask @andreasn or @garrett for design review.

andreasn commented:
Thanks for looking into this!
I wasn't able to test this locally, so will comment base on the screencasts:
* Once the report is done via the "Report to Fedora" button, the button still says "Report to Fedora", even though the button actually doesn't do that any more. Instead it's now a link to the bug report. So I think it would be clearer if it looked like a link.
* The flow for authentication in feels a bit clunky to me. The administrators goal is to just file that crash report (and maybe is pissed off already that the OS provider published something that crashes), and the UI keeps throwing obstacles in the way of actually reporting it. I think we should cut it down to only one dialog for the authentication. The reference to some configuration is a bit vague, and it's unclear how to actually fix the configuration. So maybe leave that part out for now?
* The dialog for duplicate crash has a bug with a URL, since we're in a web browser already, might be best to just make that a link that can be clicked (and the browser will show the full URL anyway).
* The Yes/No buttons in the duplicate crash dialog would be clearer if they were verbs describing the action you're taking.

![mockup](https://raw.githubusercontent.com/cockpit-project/cockpit-design/6d3cbbe0cfc8fd8122ad77ff3d677a5a7892aa62/crashes/crashes-reportd.png)

ernestask commented:
It looks completely different now, so not much point in dissecting the screencasts.

ernestask commented:
The button labels and prompts are all part of reporting workflows. If there is any problem, then ABRT itself needs design work.

andreasn commented:
So we have no controls over the labels at all?
I'm happy to file bugs against ABRT if so.

andreasn commented:
And sorry for only doing review based on the screencasts. Didn't realize they were out of date.
I'll build and test it locally instead.

marusak commented:
Hey @andreasn to ease the process of trying to crash something, here are some screenshots with description. Sorry, no video.

So when you go to a problem you see this report button:
![first](https://user-images.githubusercontent.com/12330670/57874470-f19e3d00-7810-11e9-9a4f-851f2fe50bb7.png)

When you open it, it shows workflows - there can be any number from one to n:
![workflows](https://user-images.githubusercontent.com/12330670/57874489-febb2c00-7810-11e9-8c13-2b638a0c31d3.png)

The first one is special - it is the same reporting as we have there now - it just shows spinner and then says finished.
![beforespinnerdonefaf](https://user-images.githubusercontent.com/12330670/57874550-1db9be00-7811-11e9-8798-485226532d09.png)

And then the menu looks like this (upon clicking new tab is opened)
![Screenshot from 2019-05-16 19-18-01](https://user-images.githubusercontent.com/12330670/57874593-3924c900-7811-11e9-9e40-1b9219c17508.png)

And yeah, the spinner looks like this:
![spinnertoolong](https://user-images.githubusercontent.com/12330670/57874646-56f22e00-7811-11e9-937e-a5b9ddd597ee.png)

I was "lucky" that I lost internet and I got this error - at least we can see how it looks like when it fails:
![failed](https://user-images.githubusercontent.com/12330670/57874658-5fe2ff80-7811-11e9-9111-c983b5a668c0.png)

The usual workflow would start something like this:
![firststep](https://user-images.githubusercontent.com/12330670/57874692-71c4a280-7811-11e9-9b15-7b0344510205.png)

And then there is gonna be bunch of statuses - these all come from ABRT, reportd just sends it over dbus
![uloadingstatus](https://user-images.githubusercontent.com/12330670/57874712-83a64580-7811-11e9-90bf-c9f56c32fd5a.png)

The popup can be always closed and reponed - here is on hover tooltip (also notice the spinner in button in the background):
![Screenshot from 2019-05-16 19-19-48](https://user-images.githubusercontent.com/12330670/57874760-9fa9e700-7811-11e9-92dd-4d33c7e69761.png)

When I closed the modal during processing (when you click on the button it opens the modal back):
![spinnerbutton](https://user-images.githubusercontent.com/12330670/57874834-bfd9a600-7811-11e9-99e9-f9f74b06cd79.png)

It can ask you for some input, for example for BZ account (and on other page for password, this is kinda stupid but it still is just coming from abrt and how it works)
![loginnamebz](https://user-images.githubusercontent.com/12330670/57874877-d97aed80-7811-11e9-89af-1211aa62eec6.png)

Upon finishing it looks like this - it doesn't look great but thought you could have some ideas :)
![donereportfedora](https://user-images.githubusercontent.com/12330670/57874983-1646e480-7812-11e9-9311-c48ad1c5eba8.png)

And every successfully finished workflow just adds this tick and upon clicking it is some URL:
![tworeported](https://user-images.githubusercontent.com/12330670/57875026-31b1ef80-7812-11e9-8e9a-6339944bebfd.png)

marusak commented:
To run it you need `reportd-0.6.7-1` (in Fedora in testing for both F29 and F30) and `libreport-2.10.0-3` (stable in F29 and F30). And of course `abrt-desktop` whatever version is now stable. Make sure that `abrtd`, `abrt-journal-core` and `reportd` services are running.

To create crash: `sleep 100m &` followed by `kill -SIGSEGV $!`. Then in logs you should see `Process 13081 (sleep) crashed in __nanosleep()` and thats what you are looking for.

garrett commented:
Quickly jumping in with a few comments:

- Wow, there are a lot of dialogs.
   - Are all of these separate or is there supposed to be a wizard flow?
   - Can we consolidate a bunch of these?
- "Report to" dialog is odd compared to all other dialogs in Cockpit
   - The dialog doesn't have context
   - It seems redundant to say "Report to" for all the actions. 
   - How many items will it typically have? 1, 2, or 3? Is it basically ABRT & OS vendor in most cases?
- None of the actions are shown as default actions to be done. (They're not those blue buttons.)
- "Back" is an odd action.
   - Is this from a back/forward wizard?
   - If the state is done/complate/finished, should it be "Close" (to close the dialog)?
- "Remember choice" should probably be a checkbox item.
- We shouldn't say "core dump" in the UI. It should be "crash data" (or probably something similar).
- Yes/No should not be used for dialogs. It's not clear. Make the buttons actions. Perhaps one of these button pairings:
  - "Do Not Upload" & "Upload Crash Data" (default)
  - "Do Not Send" & "Send Crash Data" (default)
- Most of the text can be improved quite a bit. Especially the "core dump" dialog text, as it's quite clunky as-is. Can we change it?
- The success, error, and uploading states are a bit odd. 
   - They should be visually clearer if something worked (green circled check icon) or not (red failed icon), or if something is pending. 
   - Having a left aligned spinner with no text is weird. A non-contextual indeterminate status spinner should have text and be centered.
   - Also, the uploading size could be improved. And do we have any way to know the total size that is to be uploaded? It's around the same size of the core dump, right? Or what? If we can get the approximate total data size, we could show a progress bar if it will take a while. (It doesn't have to be exact, as long as we don't go > 100%.)
- The "open link to report" state isn't obvious.
   - Are you able to report again? Has it already been reported? You can't tell from the text.
   - Links should look like links.
   - The circle-check complete icon should have text.
   - Is it possible to integrate the reported state and the links in the log overview instead of just being inside this dialog?

garrett commented:
Oh, I noticed the tooltip on https://user-images.githubusercontent.com/12330670/57874760-9fa9e700-7811-11e9-92dd-4d33c7e69761.png mentioning that the progress dialog can be closed.

Backing up to look at the larger picture and rethinking the whole flow:

What if we refactored the entire workflow to just not use dialogs? What if it was all inline at the top of the overview tab?

The overview could have the basic details of what went wrong and the status of reporting (not reported, currently reporting, and reported). It should be possible to report the bug with a click, see one single dialog for submission confirmation, with an opt-out for crash data (core dump) uploading. You'd be returned to the overview tab with the process. When it's complete, it has links (that look like links).

It would be great to submit reports to _all_ bug trackers at the same time too. (Perhaps a list with checkbox items for every bug report system, when bug reporting systems are > 1. If there's just one bug reporting item, it would not be a checkbox list; it would just show that it would upload to that one host.)

Doing all of this would get it down to 1 single dialog for confirmation.

BTW: We'd probably want to move all the extended details out of that tab into other tabs and only show actual high-level overview information on the tab. But that's a good idea anyway, as the "overview" isn't an overview, but a list of details.

ernestask commented:
Again, all the prompts come directly from ABRT, so, if it says that the prompt is “yes/no”, then you can only do “yes” and “no”. Same with prompt texts, same with workflow names. Not a lot of freedom to customize things.

Not using dialogs might be better, but I couldn’t come up with something that would work and look passable inline. Using check buttons to select workflows sounds great (the typical number of which would be two or three, excluding ABRT Analytics, which is a special hardcoded entry). You still have to consider the fact that the prompts will come regardless, so now you also need to differentiate between sources. Locking down the reporting to a single workflow avoided just that.

garrett commented:
I've been thinking about how we could rework crash reporting based on all of the above and have the first draft of how a revised log details page could look like — and how it could include crash reporting.

Here's a sneak peek:

![journals-details-crash](https://user-images.githubusercontent.com/10246/58431580-c3411d00-80ae-11e9-9a2a-0d437f1fdcab.png)

I have dialogs in mind, and loosely scribbled on paper, but don't have them properly implemented in a mockup yet. They're hinted at in this mockup and my comments above.

The big changes:

1. Make the breadcrumb more informational
2. Display basic information at a glance
3. Show crash reporting as a section on the page (when applicable)
4. Present an overview with critical information (from both systemd journal and the crash info)
5. Include all the extended information we currently show in a basic tabbed notebook at the bottom of the page

A diff from what we have:
1 is modified, 2 - 4 is new (although based on what we currently have, but retooled), 5 is what we currently have on the page.

I think we do need to extend the page a bit more, but can probably punt the overview and even the at-a-glance information to another PR. I can trim back the changes in another mockup so we can focus on an interim design that would be relevant to this PR. :wink:

ernestask commented:
Seems that it would increase the amount of work needed if we went for an interim design for this. I honestly don’t mind helping out with implementing a new design before fitting this in.

I do like the general idea.

ernestask commented:
Any progress on mockups?

garrett commented:
Sorry for the delay; I've been busy with #11987 and need to finish #12196 — but we're nearly there.

Meanwhile, I made a simpler version:

![journals-details-crash-simple](https://user-images.githubusercontent.com/10246/60510753-84673e00-9cd0-11e9-943c-5e7e9b4685ee.png)

(This just adds the reporting section.)

But didn't get time to do the dialogs yet. I can work on it after the above mentioned PRs land. (The first one did already and the second one isn't as complex.)

ernestask commented:
I talked with Matej just now and he thought that it might be possible to somehow do away with dialogs, since we have a well-defined set of prompts/feedback that we can get from reportd/ABRT:

- General logging,
- errors,
- alerts (any messages that require immediate attention, but not quite errors),
- free-form text input,
- yes/no questions,
- yes/no/always questions (the “always” bit could be dropped, because otherwise it’s not configurable),
- yes/no/always/never questions (same as above),
- password input.

If you think it’s a rather weird selection, then you are not alone, but it would be a tad difficult to dismantle everything right now.

ernestask commented:
What I have right now (depends on reportd 0.7.0 which is still yet to be released, but will try to expedite the process):

![Screenshot from 2019-07-16 17-09-40](https://user-images.githubusercontent.com/26867637/61306281-9971c000-a7ec-11e9-9b41-0cc00ad61330.png)
![Screenshot from 2019-07-16 17-09-46](https://user-images.githubusercontent.com/26867637/61306295-9bd41a00-a7ec-11e9-8eb5-b01e6819c91d.png)
![Screenshot from 2019-07-16 17-09-51](https://user-images.githubusercontent.com/26867637/61306297-9bd41a00-a7ec-11e9-97d8-da8e9b570973.png)
![Screenshot from 2019-07-16 17-09-56](https://user-images.githubusercontent.com/26867637/61306299-9bd41a00-a7ec-11e9-87d1-fd3955b9bc18.png)

Did away with the ability to remember responses to certain prompts, meaning that they will be shown unconditionally, because that cannot be easily reverted, so better to just annoy users more until we figure out if we want to do it and how to do it.

ernestask commented:
For multiple reports (should there ever be such a situation), I do “View reports: 1, 2, …”, where numbers are links (with the external-link icon to make them bigger).

Another thing to mention is the “ABRT Analytics” “workflow”. It’s completely custom than how the rest of it works, in that it runs a hardcoded command, which doesn’t exactly work for unprivileged users. I could try to work that into ABRT so that it is dynamic like the rest, but would graying it out/not showing it in such cases be a possibility for now?

ernestask commented:
FWIW, https://bodhi.fedoraproject.org/updates/FEDORA-2019-977681593b is in stable already, so it should be possible for others to test the changes.

garrett commented:
@ernestask: This is coming along great! Thanks work working on it.

I think we need to make the UI more like my mockup* (but adapted to the new PF4 look). The big thing here is that there should be separate sections, instead of showing everything together like it currently is in your screenshots.

_* I'm referring to this mockup, by-the-way_:

![previous mockup](https://user-images.githubusercontent.com/10246/60510753-84673e00-9cd0-11e9-943c-5e7e9b4685ee.png)

How much control do we have over the dialog in your screenshot? It's lacking a title and an appropriate action button. ("Submit" is not appropriate. It should be "Save" if it is remembered or "Send" if it's being sent over the network.) The text in that dialog needs work too, if we can (hopefully) change it.

Otherwise, I really like the progress on what I see in the screenshots! I'll have to figure out how to get this running to test it out. (I just installed Silverblue here, so it might be a little tricky.)

ernestask commented:
> How much control do we have over the dialog in your screenshot? It's lacking a title and an appropriate action button. ("Submit" is not appropriate. It should be "Save" if it is remembered or "Send" if it's being sent over the network.) The text in that dialog needs work too, if we can (hopefully) change it.

That is the most troublesome part. Without rewriting a good chunk of ABRT and the surrounding stack, the best we can do is send a generic message over the write and differentiate between types of input required (plain text, password, yes/no, that kind of deal). Using “Send” would work in all cases IMO, as I can’t imagine off the top of my head a case in which the input isn’t sent somewhere.

As for the title, yes, I intended to put either the “Report to X” string there or just the X. Completely forgot, though. If you have a better suggestion, it would be more than welcome.

I can do the sectioning, that’s no problem. Just wanted some feedback.

ernestask commented:
Oh, and speaking of action buttons - since the mechanism is very rudimentary, I cannot tell which of the actions would be the “recommended” one, so realistically I could only use a primary button in one of those free-form/password input dialogs. If that’s what you meant by “appropriate action button”, anyway.

ernestask commented:
I don’t use Silverblue, so I’m not sure if ABRT is there OOTB, but maybe it would all work just as well if you installed

- abrt
- abrt-addon-ccpp
- abrt-gui
- libreport-plugin-bugzilla
- libreport-plugin-systemd-journal
- reportd
- will-crash

and ran “will_segfault” in a terminal. You should see a notification pop up in the shell somewhere (depending on the DE) if everything went well and the crash was caught. If not, you might need to enable abrt-journal-core.service and abrtd.service.

reportd is bus-activatable, so it should just start working (in my testing, it’s fairly stable, apart from when you start doing weird things and it just dies, but it gets back up again when you reload the page).

I’m usually around on freenode as ernestask, so feel free to reach out if anything. Or ask Matej to swing by my desk to kick me.

ernestask commented:
Current iteration: ![Screenshot from 2019-07-24 16-24-02](https://user-images.githubusercontent.com/26867637/61801684-7cae3b80-ae2f-11e9-8e5c-9d1dac1b7301.png)

Will clean up and push in a minute.

garrett commented:
@ernestask: This is looking awesome!

Yesterday evening, I successfully overlaid those RPMs using `rpm-ostree install` (and then the list of packages). After a reboot and switching to your branch, it worked!

I got a bit sidetracked today, but I'll check out your branch _right now_. (It's the end of the day, so I can't promise I can do much more than this at the moment.)

One question — what's the trashcan icon for? When would someone want to delete this information? (I guess it was probably already there. Hmm.)

ernestask commented:
> Yesterday evening, I successfully overlaid those RPMs using `rpm-ostree install` (and then the list of packages). After a reboot and switching to your branch, it worked!

Whew, I was afraid it wouldn’t.

> I got a bit sidetracked today, but I'll check out your branch _right now_. (It's the end of the day, so I can't promise I can do much more than this at the moment.)

There is no rush, it’s end-of-the-day-ish here and I have other things on my plate as well.

> One question — what's the trashcan icon for? When would someone want to delete this information? (I guess it was probably already there. Hmm.)

Good catch, I was completely numb to its existence. It was there before, indeed, but I’m actually not completely sure how useful it is in Cockpit. The only thing that comes to mind is saving storage, but ABRT cleans problem directories up after some limit has been reached (on default setups).

marusak commented:
> Good catch, I was completely numb to its existence. It was there before, indeed, but I’m actually not completely sure how useful it is in Cockpit. The only thing that comes to mind is saving storage, but ABRT cleans problem directories up after some limit has been reached (on default setups).

_(before someone does `git blame` and realizes it was me who implemented that)_
There is no real strong idea why it should be there or how it is useful for Cockpit users. It just enables user to delete this ABRT report and not to show it anymore. Saving disk space would be a valid reason (some reports can be in GBs) but I think more valid usecase from cockpit would be to remove it as it is not a problem or it was reported/fixed/processed in any other way and there is no need to show it anymore.
No strong preferences if it should or shouldn't be there, though.

ernestask commented:
Hi, Garrett, did you get a chance to test the changes?

marusak commented on pkg/systemd/reporting.jsx:
```diff
@@ -0,0 +1,434 @@
+                            if ("URL" in report[1]) {
+                                reportLinks.push(report[1].URL.v.v);
+                            }
+                            this.setState({ problemState: ProblemState.REPORTED });
```
Can you set state just once at the end?

marusak commented on pkg/systemd/reporting.jsx:
```diff
@@ -0,0 +1,434 @@
+        super(props);
+
+        this.state = {
+            label: "ABRT Analytics",
```
this is not used

marusak commented on pkg/systemd/reporting.jsx:
```diff
@@ -0,0 +1,434 @@
+        }
+
+        this.setState({
+            message: "Waiting to start…",
```
This needs to be translated.

marusak commented on pkg/systemd/reporting.jsx:
```diff
@@ -0,0 +1,434 @@
+                            task_proxy.addEventListener("changed", (event, data) => {
+                                switch (data.Status) {
+                                case TaskState.CANCELED:
+                                    this.setState({ message: "Reporting was canceled", });
```
Also needs to be translated.

marusak commented on pkg/systemd/reporting.jsx:
```diff
@@ -0,0 +1,434 @@
+                                this.updateStatusFromBus();
+                            });
+                            task_proxy.addEventListener("Prompt", (event, object_path, message, type) => {
+                                this.setState({ message: "Waiting for input…" });
```
you guessed it ;)

marusak commented on pkg/systemd/reporting.jsx:
```diff
@@ -0,0 +1,434 @@
+                                    );
+                                    footerProps.actions.push(
+                                        {
+                                            caption: "Send",
```
Translation (also on other actions)

marusak commented on pkg/systemd/reporting.jsx:
```diff
@@ -0,0 +1,434 @@
+                                            style: "primary",
+                                        }
+                                    );
+                                    break;
```
are `ASK_YES_NO_YESFOREVER` and `ASK_YES_NO_SAVE` handled somehow?

ernestask replied:
No (at least not currently). I decided to not expose the “remember choice” functionality, since it might need additional bells and whistles to manage the stored configuration.

marusak commented on pkg/systemd/reporting.jsx:
```diff
@@ -0,0 +1,434 @@
+        if (props.reportLinks.length === 1) {
+            status = (
+                <a href={props.reportLinks[0]} rel="noopener noreferrer" target="_blank">
+                    View report <i className="fa fa-external-link" aria-hidden="true" />
```
This also needs translation (as well as all other statuses and all action button texts)

marusak commented on pkg/systemd/reporting.jsx:
```diff
@@ -0,0 +1,434 @@
+    const reportButton = (
+        <button className="btn btn-primary"
+                disabled={props.problemState !== ProblemState.REPORTABLE}
+                onClick={props.problemState === ProblemState.REPORTABLE && props.onReportButtonClick}>
```
If you used to conditionally omit it with onClick={condition && value}, pass onClick={condition ? value : undefined} instead.

marusak replied:
minimally to keep react from complaining :D 
But I guess they have some reason why this check exists

ernestask replied:
Again, treat me like I’m stupid. Complaining about what?

ernestask replied:
I have a feeling that these are the warnings you mentioned, but I just don’t see them popping up.

marusak replied:
if you run `autogen.sh`  with `--enable-debug` then you should see that in your browser console.

ernestask replied:
Aha, perfect! I was completely lost about these things.

marusak commented on pkg/systemd/reporting.jsx:
```diff
@@ -0,0 +1,434 @@
+                                this.setState({ message: "Waiting for input…" });
+                                const task_prompt = client.proxy("org.freedesktop.reportd.Task.Prompt", object_path);
+                                const props = {
+                                    body: message,
```
It should not be string but element.

ernestask replied:
How come? I can’t see anything that would suggest that.

marusak replied:
https://github.com/cockpit-project/cockpit/blob/master/pkg/lib/cockpit-components-dialog.jsx#L297

ernestask replied:
You’ll have to talk to me like I’m five. AFAIK, pure text is completely fine in DOM. If you look at how the property is used above, that certainly appears legal.

marusak replied:
It is legal for sure and it for sure works. But its type is defined as React element and you pass it String. So it is just for a good measure.
More about typechecking of props can be found at https://reactjs.org/docs/typechecking-with-proptypes.html

marusak commented on webpack.config.js:
```diff
@@ -110,6 +110,7 @@ var info = {
         ],
         "systemd/logs": [
             "systemd/logs.js",
+            "systemd/reporting.jsx",
```
and this should not be needed

ernestask commented on test/verify/check-journal:
```diff
@@ -475,36 +479,65 @@ s.send(b"PRIORITY=3\\nFOO=bar\\n")'
+        b.click(sel)
+
+        sel = ".modal .modal-footer .btn:contains('Back')"
+        b.wait_visible(sel)
```
To be fair, I don’t remember if I actually updated the tests to work with the current changes, so this might be left over from the previous iteration.

andreasn commented:
> can you please do design review? Thanks!

I think @garrett might be better suited to review this one, since it's based on his designs. But I'm happy to take a look as well.

marusak commented on po/Makefile.am:
```diff
@@ -90,7 +90,8 @@ po/cockpit.js.pot: po/POTFILES.js.in
 		--keyword=gettextCatalog.getString:1,3c --keyword=gettextCatalog.getPlural:2,3,4c \
-		--from-code=UTF-8 --directory=$(srcdir) --files-from=$^ | \
+		--from-code=UTF-8 --directory=$(srcdir) --files-from=$^ \
+		--add-comments=Translators | \
```
How did this come to life? Is it needed? If is, it definitely definitely be a different commit.

ernestask replied:
I was doing and redoing the translatable strings and some required translator comments. Without this, they aren’t extracted, but I guess I could split it off.

marusak replied:
ah cool. Please split it to different commit or even maybe a different PR (ideally with demoing on some string and documenting it (ideally in https://github.com/cockpit-project/cockpit/tree/master/pkg/playground in the translate subsection)

ernestask replied:
Made the strings simpler, this is no longer there.

marusak commented on pkg/systemd/reporting.jsx:
```diff
@@ -0,0 +1,467 @@
+        super(props);
+
+        let label = this.props.workflow[1];
+        const match = /(?:Report to) (.*)/.exec(this.props.workflow[1]);
```
Out of interest, this `Report to` is name of workflow - it would never be localized on the system?

marusak replied:
ok, it comes from reportd's `GetWorkflows` which always passes it in English, right?

ernestask replied:
Ah, crap, they are localizable. I don’t know why I thought otherwise. So this complicates things a little.

`Report to Fedora                     [ Report ]` - this looks a bit awkward now.

marusak replied:
When we use [spawn](https://cockpit-project.org/guide/latest/cockpit-spawn.html) we can set environment (LC_ALL=C) to work around this issue. Maybe something similar could be solution for this problem? Or introduce short name in workflows?
But if this is bigger problem, I guess then follow up is fine.

marusak commented:
btw, I just realized that this is very similar design-wise to how firewall is being redesigned (https://github.com/cockpit-project/cockpit/pull/12884) in the sense that there is some heading and panels underneath - if the heading would have white background it would match that design so much better.
But not a thing for this PR (it also does much more than it was supposed to), but a nice followup.

marusak commented:
Thanks, I triggered tests. Three things need some love:
1. Semphore - `./tools/test-static-code` is unhappy (it check python codestyle)
2. [testContainerProblems](https://209.132.184.41:8493/logs/pull-11150-20191017-055151-f987e516-cockpit-project-cockpit--fedora-30/log.html#35-2) - you probably forgot yo adjust code in this test
3. [testAbrtReport](https://209.132.184.41:8493/logs/pull-11150-20191017-055151-f987e516-cockpit-project-cockpit--fedora-30/log.html#211-2) - this is real failure. It may be related to timing

ernestask commented:
Fixed the first two, will look into the third one.

ernestask commented:
Okay, it *should* be fine now.

One thing was that restarting the reportd service was causing test failures due to explosions in D-Bus code:
```
org.freedesktop.reportd: couldn't get managed objects at /org/freedesktop/reportd: GDBus.Error:org.freedesktop.DBus.Error.NoReply: Remote peer disconnected
```

Which is entirely reasonable, so I just allowed such messages in the journal.

The races wrt prompt dialogs are a bit more complicated, so I resorted to sleeping for a bit.

martinpitt commented on pkg/systemd/logs.js:
```diff
@@ -414,13 +420,20 @@ $(function() {
+            const date = moment(new Date(entry.__REALTIME_TIMESTAMP / 1000));
 
             if (is_problem) {
+                crumb.text(cockpit.format("$0: crash @ $1", id, date.format("YYYY-MM-DD HH:mm:ss")));
```
"crash" needs to be i18n'ed. Also, please let's not use "@" but a proper word ("crashed at $1") and let that be part of the translation.

martinpitt commented on test/verify/check-journal:
```diff
@@ -427,26 +427,24 @@ s.send(b"PRIORITY=3\\nFOO=bar\\n")'
+        b.wait_in_text("#journal-entry-heading", "sleep")
         # details view should hide log view
         b.wait_not_visible('.cockpit-log-panel')
-        b.wait_present("#journal-entry-message:contains('crashed in %s')" % self.crash_fn)
```
Why was that removed? Can't abrt figure out the function it crashed in for the sleep command? If not, then moving from python to timeout+sleep wasn't a good move.

ernestask replied:
I don’t really recall by now, but I think it can be added back.

ernestask replied:
The ABRT argument is completely orthogonal to the issue. I had simply removed the element previously and forgot to add this line back when I added the element back.

ernestask replied:
A bit off-topic for this, but also see the naughties for ABRT.

martinpitt commented on test/verify/check-journal:
```diff
@@ -471,40 +476,49 @@ s.send(b"PRIORITY=3\\nFOO=bar\\n")'
 
-        # "Unreport" the problem to test reporting unknown problem
-        m.execute('find /var/spool/abrt -name "reported_to" | xargs rm')
+        # FIXME: Remove this once reportd-0.7.1 lands there.
```
This should happen [in the next days](https://bodhi.fedoraproject.org/updates/FEDORA-2019-1c185dd085); after that it will require a fedora-31 image rebuild. If you want to land it earlier, please add an `else:` that checks what happens with earlier reportd versions.

ernestask replied:
“Next days” is “when F31 goes GA”, because it’s frozen (and the preferred target is a no-go). There are different bugs in previous versions, so I’d rather not check for buggy behavior.

martinpitt replied:
Ack. Since 0.7.0 seems to make it into the final release, it should at least do something sensible then, or (better yet) have a versioned dependency to >= 0.7.1?

ernestask replied:
The versioned dependency sounds like a good solution. 0.6.x -> 0.7.0 was me adding the ability to cancel tasks, but then Matej noticed that the reporting never actually finishes in the usual case (final state change is never announced, but it technically does the job), so I fixed it up in .1.

ernestask replied:
So I should maybe also test cancellations.

martinpitt replied:
reportd 0.7.1 is in Fedora 31 now, so this should be updated now. But I'll put that onto the next review round, to make it a bit easier to track everything.

martinpitt commented on tools/cockpit.spec:
```diff
@@ -366,6 +366,7 @@ Requires: kexec-tools
 Recommends: polkit
 Recommends: NetworkManager-team
 Recommends: setroubleshoot-server >= 3.3.3
+Recommends: (reportd if abrt)
```
That will add the recommends for RHEL only. Please move it outside the `%if` so that it also applies to Fedora.

ernestask replied:
Oops.

martinpitt commented:
Can you please also attach a screenshot? The latest one here still uses the light nav design.

martinpitt commented on pkg/systemd/reporting.jsx:
```diff
@@ -0,0 +1,450 @@
+// Avoiding using the proxy to circumvent the GLib cache.
+function get_problem_property(problem, property) {
+    return cockpit
+            .dbus("org.freedesktop.problems", { superuser: "try" })
```
Please don't instantiate that D-Bus object multiple times. Do it once in some constructor or just at the top of the file.

Also, you can't immediately `.call` this, you need to wait for it first. `cockpit.dbus()` by itself doesn't really do/block on/wait for anything.

You also need to handle errors here or in all callers, in case the D-Bus call fails for some reason. Otherwise the page crashes.

In particular, please add a test case that checks the page without reportd (it can just rpm -e it, or systemctl mask it or similar). It should assert that the reporting UI is not visible then, and the page otherwise behaves as expected.

martinpitt commented on pkg/systemd/reporting.jsx:
```diff
@@ -0,0 +1,450 @@
+    UNREPORTABLE: 3,
+});
+
+// Avoiding using the proxy to circumvent the GLib cache.
```
This needs a more detailled explanation. A dbus proxy does react to property updates. However, if you only need to get one or two properties from each object, then a proxy is indeed rather heavyweight, and a raw cockpit.dbus() is more efficient. That would be a legit reason, but please adjust the comment accordingly if that's it.

ernestask replied:
It’s been months since I last looked into this, but I believe the change notifications weren’t coming and the proxy still held old values. I tracked it down to somewhere in GLib, but couldn’t tell you anything concrete at this point.

martinpitt commented on pkg/systemd/reporting.jsx:
```diff
@@ -0,0 +1,450 @@
+
+    onReportButtonClick(event) {
+        if (this.state.problemState === ProblemState.REPORTING) {
+            return;
```
This is weird -- why wouldn't the button simply be disabled while the reporting is ongoing?

ernestask replied:
Sure, why not.

martinpitt commented on pkg/systemd/reporting.jsx:
```diff
@@ -0,0 +1,450 @@
+                .done(() => {
+                    this.setState({ problemState: ProblemState.REPORTED, });
+                })
+                .fail((ex) => {
```
Please at least `console.warn()` the exception, so that we have a chance to debug problems with the reporting.

ernestask replied:
It’s not really an exception, though: https://cockpit-project.org/guide/latest/cockpit-spawn.html#cockpit-spawn-fail

ernestask replied:
I can add some output for when it’s killed or exits with a non-zero code, but the latter should be more or less reflected in the status column with a line from stderr.

ernestask commented on test/verify/check-journal:
```diff
@@ -458,6 +456,13 @@ s.send(b"PRIORITY=3\\nFOO=bar\\n")'
         b = self.browser
         m = self.machine
 
+        # We restart the reportd service, which causes the D-Bus proxy to go insane.
```
Well, the error in the journal seemed to disappear when I used the `track` option, but then it reappeared. So either I suffered temporary blindness (which happens often) or that has no effect. The “insanity” is just that, the error.

martinpitt commented on pkg/systemd/reporting.jsx:
```diff
@@ -0,0 +1,450 @@
+                                      })
+                .always(() => {
+                    this.updateStatusFromBus();
+                })
```
General style nitpick: Please avoid the unnecessary { ;} for arrow functions. These are primarily meant to be an expression, i. e. use `.always(() => this.updateStatusFromBus())`. Same issue with a lot of similar handlers in this PR.

ernestask replied:
Sure thing.

ernestask replied:
After a cursory glance, I don’t think any of the uses in the file actually are meant to be expressions, so, while that might be the general case in Cockpit, it doesn’t quite make sense to be returning values like that.

ernestask replied:
Okay, I can definitely convert *some* of them, but then readability drops because of the deep nesting (which is another problem).

martinpitt replied:
Sure it's a trade-off between "lots of extra punctuation" ( { ; }) and "more stuff on the same line. This is just a nitpick though (I generally prefer more compactness and less boilerplate), but feel free to ignore. This is by far the least important review comment here. :grinning:

martinpitt commented on pkg/systemd/reporting.jsx:
```diff
@@ -0,0 +1,450 @@
+    onReportButtonClick(event) {
+        if (this.state.problemState === ProblemState.REPORTING) {
+            return;
+        }
```
Same issue as above -- please disable buttons while there is an ongoing action, rather than ignoring the click.

martinpitt commented on pkg/systemd/reporting.jsx:
```diff
@@ -0,0 +1,450 @@
+                            task_proxy.Start();
+                        });
+                    });
+        });
```
This also needs an error handler for a failed call().

martinpitt commented:
Doesn't build, sorry:
```
ERROR in ../pkg/systemd/logs.js
Module Error (from ../node_modules/eslint-loader/dist/cjs.js):

/build/cockpit-project/bots/make-checkout-workdir/pkg/systemd/logs.js
  28:8  error  Parsing error: Identifier 'ReactDOM' has already been declared

  26 | import { ReportingTable } from "./reporting.jsx";
  27 | 
> 28 | import ReactDOM from 'react-dom';
     |        ^
  29 | import React from 'react';
  30 | import { EmptyStatePanel } from "cockpit-components-empty-state.jsx";
  31 |
```

martinpitt commented on pkg/systemd/reporting.jsx:
```diff
@@ -0,0 +1,463 @@
+                .promise()
+                .then((reportable) => {
+                    if (!reportable) {
+                        this.setState({ problemState: ProblemState.UNREPORTABLE });
```
This and line 102 look like a race condition. Both properties are being queried in parallel (with two separate D-Bus calls), so depending on which one wins, problemState may end up as UNREPORTABLE or REPORTED. It may be better to do a `.GetAll()` call and look at both properties in the same function, to avoid races and ambiguity?

ernestask replied:
That is actually very unlikely to happen, but maybe fetching everything at once would be better, indeed.

martinpitt commented on pkg/systemd/reporting.jsx:
```diff
@@ -0,0 +1,463 @@
+    //
+    // TODO: Only use system bus when reportd is merged into ABRT
+    //       (https://github.com/abrt/reportd/issues/8).
+    const bus_type = permission.allowed ? "system" : "session";
```
That PR didn't land yet, does that block landing this PR? Does it work to call it on the session bus as non-root, and does it work to call it on the system bus without that PR?

ernestask replied:
Actually, the work on that is sort of blocked by this, since I can iterate faster with reportd being a separate project, but since I more or less imagine what the requirements are, I can go ahead with that soon™.

It does work if a non-root user speaks to it over the session bus, but the main point of the comment is that the ABRT daemon is only reachable via system bus.

martinpitt replied:
Ah, I see! Sorry, I previously misunderstood the comment before.

martinpitt commented on test/verify/check-journal:
```diff
@@ -458,6 +457,15 @@ s.send(b"PRIORITY=3\\nFOO=bar\\n")'
+        self.allow_journal_messages('.*Remote peer disconnected')
+
+        m.upload(["verify/files/mock-bugzilla-server.py"], "/tmp/")
+        m.execute("setsid /tmp/mock-bugzilla-server.py >/tmp/mock-bugzilla.log 2>&1 &")
```
There's `m.spawn()` for this, which gives you a pid back. [Example](https://github.com/cockpit-project/cockpit/blob/master/test/verify/check-metrics#L106)

ernestask replied:
I’m going to assume that you are referring to the `timeout` hack for killing `sleep`.

martinpitt replied:
No, I mean that `m.execute("setsid ...")` is better done with `m.spawn("...")`. But it's not that important, I was just pointing out that this exists. The crashing sleep is fine.

ernestask commented:
Given how inconsistent the static checking is, I’m not surprised.

martinpitt commented on test/verify/check-journal:
```diff
@@ -471,41 +479,117 @@ s.send(b"PRIORITY=3\\nFOO=bar\\n")'
+            sel = "table.reporting-table tr:nth-child(2) button:contains('Report')"
+            b.click(sel)
+
+            test_user = 'correcthorsebatterystaple'
```
:+1: This is being used both as user name and password? It's just against the mock, and I suppose that just ignores it?

ernestask replied:
Yes and yes. Thought I would use something less tired, but didn’t quite feel like using unique username and password.

martinpitt commented on test/verify/check-journal:
```diff
@@ -471,41 +479,117 @@ s.send(b"PRIORITY=3\\nFOO=bar\\n")'
+
+    @skipImage("ABRT does not work on i386", "fedora-i386")
+    @skipImage("ABRT not available", "debian-stable", "debian-testing", "ubuntu-stable",
+               "ubuntu-1804", "fedora-atomic", "rhel-8-1", "rhel-8-1-distropkg")
```
This needs "rhel-8-2" as well now.

martinpitt commented:
Functionally this looks great now, thanks! Screenshot for release notes:

![logs-crash-reportd](https://user-images.githubusercontent.com/200109/68296630-5d14e900-0095-11ea-9b53-887408c6f678.png)

ernestask commented on pkg/systemd/reporting.jsx:
```diff
@@ -0,0 +1,463 @@
+                      "org.freedesktop.DBus.Properties",
+                      "Get", ["org.freedesktop.Problems2.Entry", property])
+                .done((property) => promise.resolve(property))
+                .fail((ex) => console.error(cockpit.format("Getting property $0 for problem $1 failed: $2", property, problem, ex)));
```
All right. The reason for using the ancient API is just lack of education.

martinpitt replied:
No worries -- I'm happy to port it to Promise (unless you want to, of course), fine to let this land with defers for now.

ernestask replied:
I did a quick conversion to JS promises.

ernestask commented:
Rebased and pushed with promise-related refactoring. I wanted to wait until I could create an update for reportd, but we have a bit of a blocker in libreport, so YMMV as far as testing goes (https://github.com/abrt/reportd/commit/e5b923089e84327c9593666b527c522d3a5e629f).

ernestask commented:
https://bodhi.fedoraproject.org/updates/FEDORA-2019-c5b03222c0

Unless there are more breaking changes in libreport (I’m sure there will be in the future), this should be the last of it.

martinpitt commented:
Thanks @ernestask! This has a conflict and thus needs rebasing, and it can't be built due to a trivial pyflakes error. You should be able to reproduce with just running tools/test-static code:
```
FAIL: tools/test-static-code 2 pyflakes test
test/verify/check-journal:402: local variable 'm' is assigned to but never used
test/verify/check-journal:442: local variable 'm' is assigned to but never used
```
I'll fix that locally and build your branch without rebasing for the next review round.

martinpitt commented on test/verify/check-journal:
```diff
@@ -467,54 +478,136 @@ s.send(b"PRIORITY=3\\nFOO=bar\\n")'
-        self.allow_journal_messages('.*https://bugzilla.example.com/show_bug.cgi\?id=123456')
+    @skipImage("ABRT does not work on i386", "fedora-i386")
+    @skipImage("ABRT not available", "debian-stable", "debian-testing", "ubuntu-stable",
+               "ubuntu-1804", "fedora-atomic", "rhel-8-1", "rhel-8-1-distropkg", "rhel-8-2")
```
This needs fedora-atomic → fedora-coreos, similar to the other testAbrt* in master now.

martinpitt commented on tools/cockpit.spec:
```diff
@@ -369,6 +369,11 @@ Recommends: setroubleshoot-server >= 3.3.3
+%if 0%{?fedora} >= 29
+# 0.7.0 (actually) supports task cancellation.
+# 0.7.1 fixes tasks never announcing completion.
+Recommends: (reportd >= 0.7.1 if abrt)
```
Is this ever aimed at RHEL as well, or will this stay a Fedora-ism for the time being?

ernestask replied:
For the time being, it will stay a Fedora-ism. The closest to RHEL is the EPEL 8 branch I have, but I’ll need to maintain a separate codebase.

martinpitt commented on pkg/systemd/reporting.jsx:
```diff
@@ -0,0 +1,499 @@
+
+    updateStatusFromBus() {
+        get_problem_properties(this.props.problem)
+                .catch(exception => console.error(cockpit.format("Getting properties for problem $0 failed: $1", this.props.problem.path, exception)))
```
Shouldn't this also set the state to UNREPORTABLE?

ernestask replied:
Yeah, I think that would make sense.

martinpitt commented on test/verify/check-journal:
```diff
@@ -455,54 +464,136 @@ ExecStart=/bin/sh -c 'for s in $(seq 10); do echo SLOW; sleep 0.1; done; sleep 1
+        sel = "table.reporting-table tr:nth-child(2) td:contains('Reporting was canceled')"
+        b.wait_present(sel)
+
+    def testAbrtReportNoReportd(self):
```
This test case needs to be skipped on OSes without ABRT, like here: https://logs-https-cockpit.apps.ci.centos.org/logs/pull-11150-20191118-131111-9511f328-cockpit-project-cockpit--debian-testing/log.html#253-2

ernestask replied:
Argh, completely forgot about it.

ernestask commented:
Pushed with changes.

martinpitt commented:
@ernestask , can you please push again? Doesn't build:

```
  305:43  error  Parsing error: Unexpected token

  303 | 
  304 |                         console.error(cockpit.format("reportd task for workflow $0 did not finish: $1", this.props.workflow[0], (ex.problem || ex.message)));
> 305 |                         this.setState({ _("Reporting failed") });
      |                                           ^
```

ernestask commented:
Done, sorry.
